Smt/Solver.lean

```diff
@@ -54,100 +56,157 @@ register_option smt.solver.path : String := {
 
 /-- Define a sort with name `id` and arity `n`. -/
-def declareSort (id : String) (n : Nat) : Solver :=
-  ⟨s!"(declare-sort {quoteSymbol id} {n})" :: solver.commands⟩
+def declareSort (id : String) (n : Nat) : SolverT m Unit := emitCommand (.declareSort id n)
 
 /-- Define a sort with name `id`. -/
-def defineSort (id : String) (ss : List Term) (s : Term) : Solver :=
-  ⟨(s!"(define-sort {quoteSymbol id} ({paramsToString ss}) {s})") :: solver.commands⟩
-  where
-    paramsToString (ss) := String.intercalate " " (ss.map Term.toString)
+def defineSort (id : String) (ss : List Term) (s : Term) : SolverT m Unit :=
+  emitCommand (.defineSort id ss s)
 
 /-- Declare a symbolic constant with name `id` and sort `s`. -/
-def declareConst (id : String) (s : Term) : Solver :=
-  ⟨s!"(declare-const {quoteSymbol id} {s})" :: solver.commands⟩
-
-/-- Define a constant with name `id` sort `s`, and value `v`. -/
-def defineConst (id : String) (s : Term) (v : Term) : Solver :=
-  ⟨s!"(define-const {quoteSymbol id} {s} {v})" :: solver.commands⟩
+def declareConst (id : String) (s : Term) : SolverT m Unit := emitCommand (.declare id s)
 
 /-- Declare an uninterpreted function with name `id` and sort `s`. -/
-def declareFun (id : String) (s : Term) : Solver :=
-  ⟨s!"(declare-fun {quoteSymbol id} {s})" :: solver.commands⟩
+def declareFun (id : String) (s : Term) : SolverT m Unit := emitCommand (.declare id s)
 
 /-- Define a function with name `id`, parameters `ps`, co-domain `s`,
     and body `t`. -/
 def defineFun (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
-  Solver :=
-  ⟨s!"(define-fun {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
-  where
-    paramsToString (ps : List (String × Term)) : String :=
-      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))
+  SolverT m Unit := emitCommand (.defineFun id ps s t false)
 
 /-- Define a recursive function with name `id`, parameters `ps`, co-domain `s`,
     and body `t`. -/
-def defineFunRec (id : String) (ps : List (String × Term)) (s : Term)
-  (t : Term) :
-  Solver :=
-  ⟨s!"(define-fun-rec {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
-  where
-    paramsToString (ps : List (String × Term)) : String :=
-      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))
+def defineFunRec (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
+  SolverT m Unit := emitCommand (.defineFun id ps s t true)
 
 /-- Assert that proposition `t` is true. -/
-def assert (t : Term) : Solver :=
-  ⟨s!"(assert {t})" :: solver.commands⟩
+def assert (t : Term) : SolverT m Unit := emitCommand (.assert t)
 
-def queryToString : String :=
-  String.intercalate "\n" ("(check-sat)\n" :: solver.commands).reverse
+/-- Check if the query given so far is satisfiable and return the result. -/
+def checkSat : SolverT m Result := do
+  emitCommand .checkSat
+  let out ← getSexp
```

Comment:
I think the "eager" approach might break with some solvers (forgot which) that don't process commands immediately on seeing a newline but rather wait for you to close `stdin` (which is why we have the `takeStdin` in `exit`) and then do everything at once. But let's worry about that later.